server/src/main/java/org/elasticsearch/search/fetch/subphase/FetchFieldsContext.java

```diff
@@ -36,21 +37,21 @@ public static FetchFieldsContext create(
     ) {
         DocumentMapper documentMapper = mapperService.documentMapper();
         if (documentMapper.sourceMapper().enabled() == false) {
-            throw new IllegalArgumentException("Unable to retrieve the requested [fields] since _source is " +
-                "disabled in the mappings for index [" + indexName + "]");
+            throw new IllegalArgumentException(
+                "Unable to retrieve the requested [fields] since _source is " + "disabled in the mappings for index [" + indexName + "]"
+            );
         }
 
-        FieldValueRetriever fieldValueRetriever = FieldValueRetriever.create(mapperService, searchLookup, fields);
-        return new FetchFieldsContext(fieldValueRetriever);
+        return new FetchFieldsContext((m, s) -> FieldValueRetriever.create(m, s, fields));
     }
 
-    private final FieldValueRetriever fieldValueRetriever;
+    private final BiFunction<MapperService, SearchLookup, FieldValueRetriever> fieldValueRetriever;
```

Comment:
Could we just store and expose `fields` on this context? It feels confusing to me to store a `FieldValueRetriever` supplier instead. Then we could take a simple approach where `FetchFieldsPhase` just constructs the `FieldValueRetriever` directly.

Reply:
I have no idea why I didn't do that before.....